Paper: EEG-to-fMRI Neuroimaging Cross Modal Synthesis in Python (#801)
scipy-conference/scipy_proceedings: papers/david_calhas/main.tex

\subsection{Introduction}\label{section:introduction}

> Neuronal activity, usually measured through electroencephalography (EEG), is related to haemodynamical acitivity, measured through functional magnetic resonance imaging (fMRI). The first captures the dynamics of the electrical field, whose source is located from the firing neurons' action potentials. In its turn, the second measures the blood supply dynamics. These two while being studied simultaneously \cite{shibasaki2008human,yu2016building,he2018spatialtemporaldo,rojas2018study,brechet2019capturing,daly2019electroencephalography,cury2020sparse,abreu2021eeg} differ in many aspects such as: temporal and spatial resolution, brain functions captured, recording and hardware cost. Recently, we have seen the contribution of multiple studies that target a mapping function able to provide a synthesized fMRI view from an EEG signal. Such a mapping could allow health care cost reductions and discoveries of new neuroscience insights on the relationship between these two modalities. Indeed, pathologies that require MRI scans diagnostics benefit from a lower cost EEG assessment, since availability of MRI hardware is very scarce \cite{ogbole2018survey}. As Python \cite{van1995python} becomes a hub for scientific development \cite{harris2020array,virtanen2020scipy,abadi2016tensorflow} we find, the need to provide open source software that provides solutions for \textit{EEG to fMRI synthesis}, urgent, in order for third party scientific contributions coming from other laboratories to coexist and health care software integration to develop for diagnostic settings. To that end, we provide a description of the open source software \href{https://pypi.org/project/eeg-to-fmri/}{EEG-to-fMRI}, which originated from an academic scientific project funded by FCT, and make publicly available a github \href{https://github.com/eeg-to-fmri/eeg-to-fmri}{repository}.

Reviewer: ```suggestion
Neuronal activity, usually measured through electroencephalography (EEG), is related to haemodynamical activity, measured through functional magnetic resonance imaging (fMRI). The first captures the dynamics of the electrical field, whose source is located from the firing neurons' action potentials. In its turn, the second measures the blood supply dynamics. These two while being studied simultaneously \cite{shibasaki2008human,yu2016building,he2018spatialtemporaldo,rojas2018study,brechet2019capturing,daly2019electroencephalography,cury2020sparse,abreu2021eeg} differ in many aspects such as: temporal and spatial resolution, brain functions captured, recording and hardware cost. Recently, we have seen the contribution of multiple studies that target a mapping function able to provide a synthesized fMRI view from an EEG signal. Such a mapping could allow health care cost reductions and discoveries of new neuroscience insights on the relationship between these two modalities. Indeed, pathologies that require MRI scans diagnostics benefit from a lower cost EEG assessment, since availability of MRI hardware is very scarce \cite{ogbole2018survey}. As Python \cite{van1995python} becomes a hub for scientific development \cite{harris2020array,virtanen2020scipy,abadi2016tensorflow} we find the need to provide open source software that provides solutions for \textit{EEG to fMRI synthesis}, urgent, in order for third party scientific contributions coming from other laboratories to coexist and health care software integration to develop for diagnostic settings. To that end, we provide a description of the open source software \href{https://pypi.org/project/eeg-to-fmri/}{EEG-to-fMRI}, which originated from an academic scientific project funded by FCT, and make publicly available a github \href{https://github.com/eeg-to-fmri/eeg-to-fmri}{repository}.
```

Reviewer: - [ ] 'haemodynamical acitivity' -> 'activity'

Reviewer: - [ ] Define FCT before you use it as an acronym. Especially if they are you funding source!

Reviewer: > Recently, we have seen the contribution of multiple studies that target a mapping function able to provide a synthesized fMRI view from an EEG signal.

You are correct of course that a so-called neural network model is really just a mathematical function that does a mapping. But I would suggest making it clearer for a broader audience that you are talking about deep neural network models here 

"Recently we have seen several studies that use deep neural network models (cite e.g. Goodfellow textbook) to learn a mapping from EEG data to and from fMRI data. These are a type of generative model (cite a relevant review here -- I don't know the models off the top of my head and you never name them)"

\subsection{Methods}\label{section:methods}

> The mapping function provided in this software is the one proposed by \cite{calhas2022eeg}. It consists on transforming the EEG from a channel by time representation to a channel by time frequency one, achieved using the short time Fourier transform \cite{allen1977short} by means of the fast Fourier transform (\href{https://docs.scipy.org/doc/scipy/reference/generated/scipy.fft.fft.html}{scipy.fft.fft}) available in the SciPy. The latter corresponds to the second step of the diagram illustrated in \ref{fig:pipeline}. In the second step, this representation is then forward through a deep neural network (implemented as a \href{https://www.tensorflow.org/api_docs/python/tf/keras/Model}{tf.keras.Model}), with contributions ranging from Resnet blocks \cite{he2016deep}, an automated machine learning framework \cite{calhas2022automatic} and Fourier features \cite{tancik2020fourier}. Ultimately, this enables the prediction of an fMRI volume associated with a $26$ second segment of an EEG recording. 

Reviewer: ```suggestion
The mapping function provided in this software is the one proposed by \cite{calhas2022eeg}. It consists on transforming the EEG from a channel by time representation to a channel by time-frequency one, achieved using the short time Fourier transform \cite{allen1977short} by means of the fast Fourier transform (\href{https://docs.scipy.org/doc/scipy/reference/generated/scipy.fft.fft.html}{scipy.fft.fft}) available in the SciPy package. The latter corresponds to the second step of the diagram illustrated in \ref{fig:pipeline}. In the second step, this representation is then forward through a deep neural network (implemented as a \href{https://www.tensorflow.org/api_docs/python/tf/keras/Model}{tf.keras.Model}), with contributions ranging from Resnet blocks \cite{he2016deep}, an automated machine learning framework \cite{calhas2022automatic} and Fourier features \cite{tancik2020fourier}. Ultimately, this enables the prediction of an fMRI volume associated with a $26$ second segment of an EEG recording. 
```
- [ ] 'time frequency one' -> 'time-frequency one'
- [ ] 'the SciPy' -> 'the SciPy package'

\begin{figure*}[t]
    \centering
    \includegraphics[width=0.8\textwidth]{figures/diagram.png}
    \caption{The pipeline of the EEG to fMRI synthesis project consists of processing EEG recordings, that are sourced from a human subject that goes through an EEG recording sesssion, then processing the channel by time signal and taking the short time Fourier transform. The latter, gives us the time frequency representation of the EEG signal. The novelty of this software is that it provides a model that given as input the EEG signal it predicts the corresponding fMRI volume associated with that segment. A video demonstration of the whole pipeline is available on \href{https://youtu.be/47uJbI0hU_I}{Youtube}.}
    \label{fig:pipeline}
\end{figure*}

\subsection{Description}\label{section:description}

> The dependencies of each component are described in an \href{https://github.com/eeg-to-fmri/eeg-to-fmri/blob/main/project_UML.pdf}{UML diagram}. Overall to install the package, the user is required to install the following dependencies: \textit{tensorflow 2.9.0}, \textit{matplotlib 3.5.3}, \textit{mne 0.23.4}, \textit{nilearn 0.7.0}, \textit{tensorflow\_probability 0.12.2}, \textit{tensorflow\_determinism 0.3.0}, and \textit{tensorflow\_addons 0.19.0}. As seen there is a high dependency in tensorflow related packages. This is due to the whole system provided being built in tensorflow, a library that enables automatic differentiation and is widely used for deep learning model development.

Reviewer: Please use `\texttt` instead of `\textit` here and throughout the rest of the paper when referring to pacakges, functions, etc., so that those names render in monospace instead of italics

\subsubsection{Package modules}\label{section:modules}

This package, as it is provided, has eight main modules:
\begin{itemize}

>     \item \textit{models}: here you will find the code that implements the models for synthesis and classification. The synthesis models are located in the \textit{synthesizers.py}, where the class \textit{EEG\_to\_fMRI} is implemented. This class defines a \textit{tf.keras.Model}, composed of two encoders, one for the EEG and another for the fMRI. Additionally, there is a decoder that maps the latent EEG representation, that is the output of the EEG encoder, to the estimate of the fMRI volume associated. The fMRI encoder is defined in the \textit{fmri\_ae.py} file. The task would not be complete without the extrapolation of the synthesis model to a classification task. To that end, two linear classifiers are provided, the \textit{ViewLatentContrastiveClassifier} and the \textit{ViewLatentLikelihoodClassifier}, corresponding to a contrastive latent loss and a cross entropy error driven classification, respectively. The extrapolation is simply made by taking the output of the neural flow that comes from the EEG;

Reviewer: ```suggestion
    \item \textit{models}: here you will find the code that implements the models for synthesis and classification. The synthesis models are located in the \textit{synthesizers.py}, where the class \textit{EEG\_to\_fMRI} is implemented. This class defines a \textit{tf.keras.Model}, composed of two encoders, one for the EEG and another for the fMRI. Additionally, there is a decoder that maps the latent EEG representation, that is the output of the EEG encoder, to the estimate of the fMRI volume associated. The fMRI encoder is defined in the \textit{fmri\_ae.py} file. The task would not be complete without the extrapolation of the synthesis model to a classification task. To that end, two linear classifiers are provided, the \textit{ViewLatentContrastiveClassifier} and the \textit{ViewLatentLikelihoodClassifier}, corresponding to a contrastive latent loss and a cross entropy error driven classification, respectively. The extrapolation is simply made by taking the output of the neural flow that comes from the EEG
```
- [ ] Please define "neural flow"
- [ ] Removed semi-colon from end of paragraph
```suggestion
    \item \textit{models}: here you will find the code that implements the models for synthesis and classification. The synthesis models are located in the \textit{synthesizers.py}, where the class \textit{EEG\_to\_fMRI} is implemented. This class defines a \textit{tf.keras.Model}, composed of two encoders, one for the EEG and another for the fMRI. Additionally, there is a decoder that maps the latent EEG representation, that is the output of the EEG encoder, to the estimate of the fMRI volume associated. The fMRI encoder is defined in the \textit{fmri\_ae.py} file. The task would not be complete without the extrapolation of the synthesis model to a classification task. To that end, two linear classifiers are provided, the \textit{ViewLatentContrastiveClassifier} and the \textit{ViewLatentLikelihoodClassifier}, corresponding to a contrastive latent loss and a cross entropy error driven classification, respectively. The extrapolation is simply made by taking the output of the neural flow that comes from the EEG.
```

>     \item \textit{layers}: this module contains the layers which compose the synthesizer models. Five main types of layers should be considered. First, the \textit{TopographicalAttention} computes a self attention mechanism in the channels dimension of the EEG \cite{calhas2022eeg}, this allows the representation to be correctly processed by the convolutional blocks. Second, the \textit{ResnetBlock} \cite{he2016deep} is a layer corresponds to the implementation of a residual block that is composed of convolutional layers. This block allows efficient gradient propagation by tackling the vanishing gradient phenomena. Third, the \textit{FourierFeatures} is a layer that projects sinusoid functions, namely \textit{cosines}, of different shifts and biases, to build a latent spectral basis for the prediction of the fMRI volume. Fourth, the \textit{DenseVariational} is an implementation of the \textit{tf.keras.layers.Dense} for two dimensional inputs, whose weights are drawn from gaussian distributions. Last, but not least, we provide the \textit{DCT} based layers, which implement the discrete cosine transform \cite{ahmed1974discrete}. These layers are useful for alternative ways of decoding the latent representation of the EEG to produce the desired fMRI volume;

Reviewer: ```suggestion
    \item \textit{layers}: this module contains the layers which compose the synthesizer models. It provides five main types of layers. First, the \textit{TopographicalAttention} computes a self attention mechanism in the channels dimension of the EEG \cite{calhas2022eeg} that allows the representation to be correctly processed by the convolutional blocks. Second, the \textit{ResnetBlock} \cite{he2016deep} implements a residual block composed of convolutional layers. This block allows efficient gradient propagation,  by tackling the vanishing gradient phenomena. Third, the \textit{FourierFeatures} layer projects \textit{cosines} functions of different shifts and biases to build a latent spectral basis for the prediction of the fMRI volume. Fourth, the \textit{DenseVariational} is an implementation of the \textit{tf.keras.layers.Dense} layer for two-dimensional inputs, whose weights are drawn from gaussian distributions. Last, but not least, we provide the \textit{DCT} based layers, which implement the discrete cosine transform \cite{ahmed1974discrete}. These layers are useful for alternative ways of decoding the latent representation of the EEG to produce the desired fMRI volume.
```
Please consider revising as suggested

Author: I really appreciate the rephrasing! Thank you!

\subsubsection{Classification}\label{section:classification}

\begin{figure*}[t]
    \centering
    \includegraphics[width=\textwidth]{figures/discriminative.png}

>     \caption{Output of the predicted fMRI when given an EEG representation. Note that, due to the EEG encoder being optimized towards classifying the data according to the groups of individuals defined, e.g. schizophrenic and healthy controls, the decoder (that has the parameters frozen) is given a slighthly altered representation. This change is seen in the produced fMRI.}

Reviewer: Please specify what exactly in the figure shows that there is a change in the produced fMRI because of how the EEG encoder is optimized.

    \label{fig:discriminative}
\end{figure*}

> We also provide a compressed version of the dataset of \cite{dataset2022fribourg}. This example is based on a public available dataset that contains individuals diagnosed with schizophrenia and healthy controls. The whole goal of the project is to be applied in an health care setting and to this end we employ an end to end software solution. The whole software package is able to synthesize fMRI and adapt to a classification setting, that given EEG recordings it outputs a set of probabilities for each group of people considered in the dataset.

Reviewer: ```suggestion
We also provide a compressed version of the dataset of \cite{dataset2022fribourg}. This example is based on a publicly available dataset that contains individuals diagnosed with schizophrenia and healthy controls. The whole goal of the project is to be applied in an health care setting and to this end we employ an end to end software solution. The whole software package is able to synthesize fMRI and adapt to a classification setting, that given EEG recordings outputs a set of probabilities for each group of people considered in the dataset.
```
- [ ] "public available" -> "publicly available"
- [ ] "given EEG recordings it" -> "given EEG recordings outputs"

%\begin{figure*}[t]
%    \centering
%    \includegraphics[width=\textwidth]{figures/scz_explanations.png}
%    \caption{}
%    \label{fig:scz_explanations}
%\end{figure*}

%\begin{figure*}[t]
%    \centering
%    \includegraphics[width=\textwidth]{figures/hc_explanations.png}
%    \caption{}
%    \label{fig:hc_explanations}
%\end{figure*}

\subsection{Collaboration}\label{section:collabs}

> Ultimately, the goal of this package is to ensemble, in one package, the scientific contributions in the area of EEG to fMRI synthesis. Works such as \cite{calhas2022eeg} and \cite{liu2019convolutional} should address their contributions and proceed with pull requests for contributions. We plan to open a module or example folder that implements these contributions, so that third party laboraties can easily access and reproduce these works. 

Reviewer: ```suggestion
Ultimately, the goal of this package is to collect, in one package, methods for EEG to fMRI synthesis. We would welcome contributions from authors of related work such as \cite{calhas2022eeg} and \cite{liu2019convolutional}. In the future, we plan to add a module or example folder with implementations of these approaches, so that other research groups can easily access them and reproduce key results. 
```
- [ ] "to ensemble" -> "to collect"
- [ ] "the scientific contributions in the area of" -> "methods for"
- [ ] see additional suggestions for re-wording

Reviewer: It's a bit confusing you say "Works such as \cite{calhas2022eeg} ... should address their contributions and proceed with pull requests for contributions" here even though you say in the intro that "The mapping function provided in this software is the one proposed by \cite{calhas2022eeg}."

Are there additional methods in your 2022 paper that are not implemented in eeg-to-fmri yet? Please specify.

Author: Initially I didn't want to target the Liu study, could maybe feel as if I was attacking them. The whole package contains the code of the calhas2022eeg and more.
